packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift

```diff
@@ -0,0 +1,248 @@
+import Foundation
+import ExpoModulesCore
+protocol OrientationListener {
+  func screenOrientationDidChange(_ orientation: UIInterfaceOrientation)
+}
+
+/**
+ This singleton that holds information about desired orientation for every app which uses expo-screen-orientation.
+ Marked @objc and public, because this it is also used in EXAppViewController.
+ */
+@objc
+public class ScreenOrientationRegistry: NSObject, UIApplicationDelegate {
+  @objc
+  public static let shared = ScreenOrientationRegistry()
+
+  var currentScreenOrientation: UIInterfaceOrientation
+  var orientationListeners: [ScreenOrientationModule?] = []
+  var moduleInterfaceMasks: [ScreenOrientationModule: UIInterfaceOrientationMask] = [:]
+  weak var currentTraitCollection: UITraitCollection?
+  var lastOrientationMask: UIInterfaceOrientationMask
+  var rootViewController: UIViewController? {
+    return UIApplication.shared.windows.filter { $0.isKeyWindow }.first?.rootViewController
+  }
+
+  var currentOrientationMask: UIInterfaceOrientationMask {
+    var currentOrientationMask: UIInterfaceOrientationMask = []
+
+    EXUtilities.performSynchronously {
+      currentOrientationMask = self.rootViewController?.supportedInterfaceOrientations ?? []
+    }
+    return currentOrientationMask
+  }
+
+  private override init() {
+    self.currentScreenOrientation = .unknown
+    self.currentTraitCollection = nil
+    self.lastOrientationMask = []
+
+    super.init()
+
+    NotificationCenter.default.addObserver(
+      self,
+      selector: #selector(self.handleDeviceOrientationChange(notification:)),
+      name: UIDevice.orientationDidChangeNotification,
+      object: UIDevice.current
+    )
+
+    // This is most likely already executed on the main thread, but we need to be sure
+    RCTExecuteOnMainQueue {
+      UIDevice.current.beginGeneratingDeviceOrientationNotifications()
+    }
+  }
+
+  /**
+   Called by ScreenOrientationAppDelegate in order to set initial interface orientation
+   */
+  func updateCurrentScreenOrientation() {
+    let windows = UIApplication.shared.windows
+    if !windows.isEmpty {
+      self.currentScreenOrientation = windows[0].windowScene?.interfaceOrientation ?? .unknown
+    }
+  }
+
+  deinit {
+    EXUtilities.performSynchronously {
+      UIDevice.current.endGeneratingDeviceOrientationNotifications()
+    }
+  }
+
+  // MARK: - Affecting screen orientation
+  /**
+   Rotates the view to currentScreenOrientation or default orientation from the orientationMask
+   */
+  func enforceDesiredDeviceOrientation(withOrientationMask orientationMask: UIInterfaceOrientationMask) {
+    var newOrientation = orientationMask.defaultOrientation()
+
+    if orientationMask.contains(currentScreenOrientation) {
+      newOrientation = currentScreenOrientation
+    }
+
+    guard newOrientation != .unknown else {
+      return
+    }
+
+    RCTExecuteOnMainQueue { [weak self] in
```

Comment:
```suggestion
    DispatchQueue.main.async { [weak self] in
```
I would use the `DispatchQueue.main` instead of `RCTExecuteOnMainQueue`. The effect is the same, but we're using fewer things from the react native, which is good.

Reply:
That was me who recommended to use `RCTExecuteOnMainQueue` here. There is a small difference – if it's currently on the main thread, the work will be executed synchronously without deferring it. I'm fine with using it from RN for now, but we should make our own utils for that.